### ISSUE TITLE
Add training command-line entry point

### ISSUE THREAD

**Issue by ryparikh** [ryparikh/AdvanceWarsServer#8]
Part of the AlphaZero-style self-play training loop work.

## Context

The training loop should consume self-play replay data, optimize the policy/value model, and write checkpoints. This starts as a simple supervised-style C++/LibTorch pass over replay samples.

The detailed design reference should live in `docs/TRAINING_COMMAND_DESIGN.md`.

## V1 Scope

- Add a top-level `-train` command.
- Require `--checkpoint-in <checkpoint-dir>`; fresh model initialization stays in `-model-init`.
- Accept `--replay <file-or-directory>` where files are `standard-gl-self-play-replay-v1` JSONL shards.
- Validate replay shards by default before training.
- In directory mode, consume `.jsonl` shards sorted by path and fail if any shard is invalid or if no shards/samples are found.
- Load compact sample records in memory, optionally cap them with `--max-samples`, and seeded-shuffle each epoch.
- Lazily reconstruct sample states from `initialState` plus action prefix during batch assembly.
- Recompute legal action indices and state tensor checksums during batch assembly and fail on mismatch.
- Train policy against normalized MCTS visit distributions over all legal actions, excluding illegal actions.
- Train value with MSE against current-player outcome, including action-limit samples with target `0`.
- Use AdamW only, constant learning rate only, and float32 training only.
- Expose `--learning-rate`, `--weight-decay`, `--policy-loss-weight`, `--value-loss-weight`, `--max-grad-norm`, `--epochs`, `--batch-size`, `--max-samples`, `--seed`, `--device`, and logging/checkpoint options.
- Train on the final partial batch rather than dropping it.
- Preserve model architecture and compatibility metadata from `--checkpoint-in`; no architecture overrides in `-train`.
- Write a final `--checkpoint-out` only after successful completion.
- Support optional periodic epoch checkpoints for interruption recovery and inspection.
- Write `training.json` beside `metadata.json` and `model.pt` with run provenance, aggregate metrics, compact per-epoch history, and `trainingMetadataVersion: 1`.
- Print human-readable startup/epoch metrics plus concise failure summaries with replay/progress/error context.
- Add reusable C++ trainer APIs with a thin CLI wrapper.
- Add focused `-test-train` coverage.

## Command Shape

```powershell
..\x64\Release\AdvanceWarsServer.exe -train `
  --replay <file-or-directory> `
  --checkpoint-in <checkpoint-dir> `
  --checkpoint-out <checkpoint-dir> `
  --epochs 1 `
  --batch-size 32 `
  [--max-samples <n>] `
  --learning-rate 0.001 `
  [--weight-decay 0.0001] `
  [--policy-loss-weight 1.0] `
  [--value-loss-weight 1.0] `
  [--max-grad-norm 0] `
  [--seed 0] `
  [--log-every-batches <n>] `
  --device auto `
  [--force] `
  [--checkpoint-every-epochs <n> --checkpoint-dir <dir>]
```

## Work Items

- [ ] Add reusable trainer API types such as `TrainingOptions`, `TrainingSummary`, `TrainingError`, and `RunTraining(...)`.
- [ ] Add `Training`, `TrainingCommand`, and `TrainingTest` source/header files.
- [ ] Add command-line parsing and strict early option/path validation.
- [ ] Load replay file or directory inputs deterministically.
- [ ] Validate replay shards before consuming samples.
- [ ] Build compact in-memory sample records, apply optional `--max-samples`, and seeded per-epoch shuffle.
- [ ] Lazily reconstruct sample states and encode `StateTensor` batches.
- [ ] Recompute legal action indices and tensor checksums during batch assembly.
- [ ] Implement sparse legal-action policy loss against MCTS visit distributions.
- [ ] Implement value MSE loss against replay outcomes.
- [ ] Implement AdamW training with optional gradient clipping.
- [ ] Save final checkpoints and optional periodic checkpoints.
- [ ] Write `training.json` with provenance, resolved options, metrics, and per-epoch history.
- [ ] Print startup, epoch, optional batch, success, and failure summaries.
- [ ] Add focused `-test-train` tests proving tiny deterministic training changes at least one model parameter.
- [ ] Document `-train` and `-test-train` in `BUILD_AND_TEST.md`.
- [ ] Add a README smoke sequence chaining `-model-init`, `-self-play`, and `-train`.
- [ ] Update `TRAINING_LOOP_WORK_ITEMS.md` after implementation.

## Additional V1 Defaults

- Empty replay inputs fail.
- Epoch shuffling is deterministic from `--seed` plus epoch number.
- No validation split, holdout metrics, early stopping, or best-checkpoint selection in #8.
- Single-process and single-worker; no background prefetch or parallel replay loading.
- Console output is human-readable; `training.json` is machine-readable.
- The model is in train mode during optimization and eval/no-grad mode for inference-only checks.
- Trainer errors use structured codes and context fields.
- #8 consumes replay data only; neural-guided self-play and PUCT priors remain out of scope.

## Deferred / Follow-up Work

- #9 owns checkpoint evaluation, win rate, map/side/CO breakdowns, promotion rules, and best-checkpoint selection.
- #180 owns streaming, bounded-buffer, reservoir, or other large-corpus replay loading modes.
- #181 owns optimizer-state resume, learning-rate schedules, scheduler state, mixed precision, scaler state, and serious long-run resumability.
- #182 owns atomic training checkpoint publish hardening.
- Existing follow-ups #172 and #174 still cover optional materialized replay caches and compressed replay shards.

## Done When

A user can initialize a model checkpoint, generate or provide validated replay data, run `-train` against that replay data, and produce a new checkpoint bundle with `metadata.json`, `model.pt`, and `training.json`. Focused tests prove the trainer parses data, computes losses, trains at least one parameter, handles partial final batches, rejects bad inputs early, and writes usable checkpoint output.

Closes #8 when implemented.
